### PR TITLE
apply X-Cli rewrite of pnio layer (#443)

### PR DESCRIPTION
**Pull request by smainand**
Hello,

With the objective of closing ProfinetIO related PR, this one takes the @X-Cli #443 PR, apply it upstream and adapt it for py3.
Breakdown of the modifications can be seen in the unsquashed version at https://github.com/smainand/scapy/tree/new_v_pnio

Seb

fixes #443

### CODE REVIEW

**Comment by gpotter2** [comment]
Thanks a lot for the initiative !

You will notice that we have added quite a lot of health checks since this PR, checking mostly PEP8 & a few grammar mistakes. Here are the results of the last test:

- PEP8
```
scapy/contrib/pnio.py:178:57: E126 continuation line over-indented for hanging indent
scapy/contrib/pnio.py:179:53: E126 continuation line over-indented for hanging indent
scapy/contrib/pnio.py:254:37: E126 continuation line over-indented for hanging indent
scapy/contrib/pnio.py:257:33: E126 continuation line over-indented for hanging indent
```

- Spelling
```
scapy/contrib/pnio.py:157: Defaut  ==> Default
```

**Comment by smainand** [comment]
Thanks, I should have checked directly with tox in the first place :)

File: scapy/contrib/pnio.uts [hunk 1]
```diff
@@ -1,8 +1,12 @@
+# coding: utf8
 % ProfinetIO layer test campaign
 
 + Syntax check
 = Import the ProfinetIO layer
 from scapy.contrib.pnio import *
+import scapy.config
+import re
+scapy.config.conf.debug_dissector=True
```
Comment (gpotter2) [inline scapy/contrib/pnio.uts]: - don't use static access
- useally we do

```
old_conf_dissector = conf.debug_dissector
conf.debug_dissector = True
[...]
conf.debug_dissector = old_conf_dissector
```

File: scapy/contrib/pnio.uts [hunk 2]
```diff
@@ -20,13 +24,202 @@ p.dport == 0x8892
 
 = Ether guesses ProfinetIO as payload class
 p = Ether(hex_bytes('ffffffffffff00000000000088920102'))
-p.payload.__class__ == ProfinetIO and p.frameID == 0x0102
+type(p.payload) == ProfinetIO and p.frameID == 0x0102
```
Comment (gpotter2) [inline scapy/contrib/pnio.uts]: I think `isinstance` is even better

File: scapy/contrib/pnio.py [hunk 3]
```diff
@@ -14,67 +15,374 @@
...
+    ]
+
+    @classmethod
+    def is_extension_set(cls, _pkt, _lst, p, _remain):
+        ret = cls if isinstance(p, type(None)) or p.extension != 0 else None
+        return ret
+
+    @classmethod
+    def get_len(cls):
+        return sum([type(fld).i2len(None, 0) for fld in cls.fields_desc])
```
Comment (gpotter2) [inline scapy/contrib/pnio.py]: You can use a generator with sum
```
sum(type(fld).i2len(None, 0) for fld in cls.fields_desc)
```

File: scapy/contrib/pnio.py [hunk 4]
```diff
@@ -14,67 +15,374 @@
...
+    0xFEFC: 'DCP-Hello-Req',
+    0xFEFD: 'DCP-Get-Set',
+    0xFEFE: 'DCP-Identify-ReqPDU',
+    0xFEFF: 'DCP-Identify-ResPDU',
+    0xFF00: 'PTCP-AnnouncePDU',
+    0xFF20: 'PTCP-FollowUpPDU',
+    0xFF40: 'PTCP-DelayReqPDU',
+    0xFF41: 'PTCP-DelayResPDU-followup',
+    0xFF42: 'PTCP-DelayFuResPDU',
+    0xFF43: 'PTCP-DelayResPDU',
```
Comment (gpotter2) [inline scapy/contrib/pnio.py]: Those changes are a bit un-necessary :/
Also, I prefer double quotes (but that's not the point)

Comment (smainand) [reply scapy/contrib/pnio.py]: Is there a coding convention for Scapy regarding ' and " ?
I just usually put single quotes everywhere as it make the code "lighter" to read (less quotes are displayed for the same result).

Anyway I'll undo it at least for the dict déclaration

Comment (gpotter2) [reply scapy/contrib/pnio.py]: Thanks !

There is no convention about that :/ Scapy is full of both

**Comment by guedou** [comment]
I use “ for all strings and ‘ for single characters like in C.

**Comment by smainand** [comment]
done

**Comment by X-Cli** [comment]
Excellent work, @smainand ! Thank you!

**Pull request merged**